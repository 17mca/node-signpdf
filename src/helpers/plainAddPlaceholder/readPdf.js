import SignPdfError from '../../SignPdfError';
import readRefTable from './readRefTable';
import findObject from './findObject';

/**
 * @param {Buffer} pdf
 */
const readPdf = (pdf) => {
    const trailerStart = pdf.lastIndexOf('trailer');
    const trailer = pdf.slice(trailerStart, pdf.length - 6);

    if (trailer.lastIndexOf('/Prev') !== -1) {
        throw new SignPdfError(
            'Incrementally updated PDFs are not yet supported.',
            SignPdfError.TYPE_PARSE,
        );
    }

    let rootSlice = trailer.slice(trailer.indexOf('/Root'));
    rootSlice = rootSlice.slice(0, rootSlice.indexOf('/', 1));
    const rootRef = rootSlice.slice(6).toString().trim(); // /Root + at least one space

    let xRefPosition = trailer.slice(trailer.lastIndexOf('startxref') + 10).toString();
    xRefPosition = parseInt(xRefPosition);
    const refTable = readRefTable(pdf, xRefPosition);

    const root = findObject(pdf, refTable, rootRef).toString();
    if (root.indexOf('AcroForm') !== -1) {
        throw new SignPdfError(
            'The document already contains a form. This is not yet supported.',
            SignPdfError.TYPE_PARSE,
        );
    }
    if (refTable.maxOffset > refTable.tableOffset) {
        throw new SignPdfError(
            'Ref table is not at the end of the document. This document can only be signed in incremental mode.',
            SignPdfError.TYPE_PARSE,
        );
    }

    return {
        xref: refTable,
        rootRef,
        root,
        trailerStart,
    };
};

export default readPdf;
