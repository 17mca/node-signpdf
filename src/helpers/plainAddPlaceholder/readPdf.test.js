import fs from 'fs';
import readPdf from './readPdf';
import SignPdfError from '../../SignPdfError';

describe('readPdf', () => {
    it('Errors when a PDF that already contains forms is passed in', () => {
        // const pdf = fs.readFileSync(`${__dirname}/../../../resources/formexample.pdf`);

        // try {
        //     readPdf(pdf);
        //     expect('here').not.toBe('here');
        // } catch (e) {
        //     expect(e instanceof SignPdfError).toBe(true);
        //     expect(e.type).toBe(SignPdfError.TYPE_PARSE);
        //     expect(e.message).toMatchSnapshot();
        // }
    });
});
