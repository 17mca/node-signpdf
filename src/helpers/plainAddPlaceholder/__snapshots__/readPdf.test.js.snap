// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readPdf Errors an incrementally updated PDF is passed in 1`] = `"Incrementally updated PDFs are not yet supported."`;
